### ISSUE TITLE
WaveInEvent blocks thread pool

### ISSUE THREAD

**Issue by markboyall** [naudio/NAudio#539]
We've had an issue in our application with thread pool starvation. We found a few places where we were starving the thread pool by queuing long-running, blocking work on it. We fixed our own instances and the problem (mostly) went away. 

I was looking at the NAudio source recently for another issue. I noticed that WaveInEvent makes exactly the same mistake we did- firing a long-running, blocking task on the thread pool. See https://theburningmonk.com/2010/03/threading-using-the-threadpool-vs-creating-your-own-threads/, https://blogs.msdn.microsoft.com/vancem/2018/10/16/diagnosing-net-core-threadpool-starvation-with-perfview-why-my-service-is-not-saturating-all-cores-or-seems-to-stall/, https://docs.microsoft.com/en-us/dotnet/standard/threading/the-managed-thread-pool for articles about this. This explains how we were able to starve a four-thread thread pool by blocking only three threads, and why the issue only occurred whilst recording.

I would imagine there are probably other places where the thread pool is starved by NAudio.

**Comment by markheath** [comment]
Yes, NAudio does use a thread on the threadpool, for WaveInEvent, but you would typically only be using one thread, so threadpool starvation would require quite a lot of other things going on in your application at the same time (probably not a good pattern for handling live audio). But I agree, should probably switch to using Task.Run with TaskCreationOptions.LongRunning in the future

**Comment by markboyall** [comment]
Unfortunately, we are still using .NET 3.5 :( We did have other thread pool starvation issues, just not enough to fully cause a problem.

**Comment by markheath** [comment]
Just revisiting and cleaning up some old NAudio issues with the help of AI. There have been a number of improvements in this area in NAudio since the original issue was raised. Closing this with a bit more info...

You were right, and thank you for the detailed write-up. At the time of this report `WaveInEvent.StartRecording` queued its long-running, blocking recording loop onto the thread pool (`ThreadPool.QueueUserWorkItem(...)`), and that loop blocks for the entire recording session — exactly the thread-pool starvation pattern you described. This has since been fixed: in NAudio 3 (currently in development on `main`) the recording loop runs on a dedicated background thread rather than a thread-pool worker, so it no longer consumes a pool thread for the duration of a capture session. As part of the same WinMM revamp the class was renamed from `WaveInEvent` to `WaveIn` (the old `WaveInEvent` name remains as an `[Obsolete]` subclass for source compatibility and inherits the fix). 

NAudio 3 also introduces a modern `WasapiRecorder` with an async capture model and its own dedicated capture thread, which is the recommended API for new capture code. Note that NAudio 3 targets .NET 9+, so it won't help your .NET 3.5 application directly — but the underlying issue is resolved going forward. Closing as fixed; thanks again for catching this.
